L46-core-tls-credential-API.md

```diff
@@ -0,0 +1,501 @@
+
+The next sections explain these interfaces in more detail.
+
+#### Custom Validation Checks on the Peer Certificate Chain
+```c++
+// Contains the information from the (verified) peer certificate chain that can
+// be used to perform custom validation checks. Users should not directly
+// create or destroy this request object, but shall interact with it through
+// CertificateVerifier's Verify() and Cancel().
+// TODO(gtcooke94) Add expected
+// formats for all string values here.
+class TlsCustomVerificationCheckRequest {
+ public:
+  // The target hostname that is expected to appear in the server leaf
+  // certicate, e.g. github.com. Empty on the server-side.
+  absl::string_view target_name() const;
+
+  // The PEM-encoded leaf cert.
+  absl::string_view peer_cert() const;
+
+  // The PEM-encoded unverified peer cert chain, ordered from leaf to root, but
+  // not including the root.
+  absl::string_view peer_cert_full_chain() const;
+
+  // The common name string from the Subject extension in the peer's leaf
+  // certificate.
+  absl::string_view common_name() const;
+
+  // The subject name of the root certificate used to verify the peer chain
+  // If verification fails or the peer certificate is self-signed, this will be an
+  // empty string. If verification is successful, it is a comma-separated list,
+  // where the entries are of the form "FIELD_ABBREVIATION=string"
+  // ex: "CN=testca,O=Internet Widgits Pty Ltd,ST=Some-State,C=AU"
+  // ex: "CN=GTS Root R1,O=Google Trust Services LLC,C=US"
+  absl::string_view verified_root_cert_subject() const;
+
+  // The (possibly empty) list of URI names from the Subject Alternative Name
+  // extension in the peer's leaf certificate.
+  std::vector<absl::string_view> uri_names() const;
+
+  // The (possibly empty) list of DNS names from the Subject Alternative Name
+  // extension in the peer's leaf certificate.
+  std::vector<absl::string_view> dns_names() const;
+
+  // The (possibly empty) list of email names from the Subject Alternative Name
+  // extension in the peer's leaf certificate.
+  std::vector<absl::string_view> email_names() const;
+
+  // The (possibly empty) list of IP names from the Subject Alternative Name
+  // extension in the peer's leaf certificate.
+  std::vector<absl::string_view> ip_names() const;
+}
+
+// The base class of all verifier implementations. Note that custom
+// verifier implementations can compose their functionality with existing
+// implementations of this interface, such as HostnameVerifier, by delegating
+// to an instance of that class.
+class CertificateVerifier {
+ public:
+  virtual ~CertificateVerifier() = default;
+
+  // Verifies a connection request. The check on each internal verifier could be
```

Comment:
I think the second sentence here is confusing, since you're referencing implementation details in an interface. Can we expand the second sentence into the following?

"""
Returns true if verification occurs synchronously, in which case the verification result is populated in the sync_status output parameter. Returns false if verification occurs asynchronously, in which case the callback must be called when verification is complete and the verification result must be populated in the sync_status output parameter. 
"""